cpp/src/parquet/encryption.h

```diff
@@ -0,0 +1,515 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_ENCRYPTION_H
+#define PARQUET_ENCRYPTION_H
+
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "parquet/exception.h"
+#include "parquet/schema.h"
+#include "parquet/types.h"
+
+namespace parquet {
+
+static constexpr ParquetCipher::type kDefaultEncryptionAlgorithm =
+    ParquetCipher::AES_GCM_V1;
+static constexpr int32_t kMaximalAadMetadataLength = 256;
+static constexpr bool kDefaultEncryptedFooter = true;
+static constexpr bool kDefaultCheckSignature = true;
+static constexpr bool kDefaultAllowPlaintextFiles = false;
+static constexpr int32_t kAadFileUniqueLength = 8;
+
+class ColumnDecryptionProperties;
+using ColumnPathToDecryptionPropertiesMap =
```

Comment:
`ColumnDecryptionProperties` already has the `schema::ColumnPath`. What is the purpose of this map?

Reply:
If the purpose it to have a fast look up, we can probably keep a sorted vector of `schema::ColumnPath` and use [binary search](http://www.cplusplus.com/reference/algorithm/binary_search/)

Reply:
yep, the idea is to create these properties once for each relevant column, and to re-use them later. This doesn't happen much though, maybe twice or so. Also, typically the number of encrypted columns in not high. In any case, this is not in the critical performance path - because the main bottlenecks are data processing, and an interaction with a remote KMS . So we thought a map would be good enough.

Reply:
The map is fine. Looking at the usage, should the key be "std::string columnPath" instead? `CmpColumnPath` is converting them to strings anyway

Reply:
ok, we can do that.

Reply:
Hi Deepak, we've started working on that, but turns out this has two sides. The API, and internal SPI, work with schema::ColumnPath objects, so this conversion happens one way or another. Fortunately, this is a low frequency operation, so no effect on performance. I hope you're ok with leaving this code as is.

Reply:
I will be at Strata NY next two days, but after I come back I want to take a look at this.